### PR TITLE
Feature/reified logs

### LINKED ISSUE #41: Method of discovering the last successful run time
In the usual case, we're going to want to query for volume changes since the last time the scheduler/enqueuer ran.  Thus, the scheduler/enqueuer (or whatever instantiates it) is going to need an easy way to access that information without having to know the format of the log.  

Thoughts below: 
----

One such method is something like this:
```ruby
RunTimeManagerOrBetterName.new(profile, log_path).last_until(Time.now) do |time_range|
  # do stuff
end
```

Our case is slightly more complex in that the `do stuff` section will be invoking the scheduler, and using its output to write the incremental log.  So there's some juggling of file handles there.  Could decide to let the RunTimeManager manage the log entirely, which it could do if the block returned saved+deleted volumes.

### PR DESCRIPTION
**Pull request by malakai97**
Resolves #41, #12, #10 

* Report: Writes a report from lists of saved/deleted volumes and a time 
* ReportSummary: Reading and writing summaries of reports, most importantly the time period covered.
* ReportManager: Manages writing reports from related runs.  Contains functionality to discover the period of the most recent report, yield a new time range, and write a new report when the block finishes.

This PR is ignorant of the functionality of VolumeActionLogger.

### CODE REVIEW

**Comment by malakai97** [comment]
Arrays and hashes are pass-by-reference in ruby, so I don't expect there to be an issue passing them around.  I do agree that your suggestion is probably an improvement.  I couldn't find a block syntax for writing files a line at a time, only for reading, but I'll admit I didn't look especially hard.

Usage would be like this:
```ruby
def perform
  Datasets::ReportManager.new(report_dir).build_next_report do |time_range|
    scheduler = Scheduler.new(
      # other params....
      time_range: time_range
    )
    [scheduler.add, scheduler.delete]
  end
end
```

**Comment by aelkiss** [comment]
Right, more concerned about having the potentially large array at all. Let me finish the test I'm doing now and then I'll see if I can mock something up for you.

**Comment by malakai97** [comment]
We'll want to get caught up by running this many times over small intervals to avoid having changelists > 1 million items.

**Comment by aelkiss** [comment]
I thought the plan was to run it the first time with the entire repository (15 million volumes)

**Comment by aelkiss** [comment]
`VolumeActionLogger` is super cheesy and the absolute minimum we would need to do, this is a substantial improvement, so fine to just dump it

**Comment by aelkiss** [comment]
OK what if we did something like

```ruby
class Report
  def self.open(time_range,report_dir,fs)
    fs.mkdir_p(report_dir)
    fs.open(report_dir + "saved.txt","w") do |saved_output|
      fs.open(report_dir + "deleted.txt","w") do |deleted_output|
	report = Report.new(saved_output,deleted_output,time_range,fs)
	yield report
	report.save(report_dir)
      end
    end
  end

  def initialize(saved_output,deleted_output,time_range,fs)
    @saved_output = saved_output
    @saved_count = 0
    @deleted_output = deleted_output
    @deleted_count = 0
    @time_range = time_range
    @fs = fs
  end

  def report_saved(volume)
    @saved_output.puts(volume_to_s(volume))
    @saved_count += 1
  end

  def report_deleted(volume)
    @deleted_output.puts(volume_to_s(volume))
    @deleted_count += 1
  end

  def save(report_dir)
    fs.write(report_dir + "summary.yml", summary.to_h.to_yml)
  end

  def summary
    ReportSummary.new(@saved_volume_count, @deleted_volume_count, time_range, fs)
  end
end
```

where `fs.open` is a wrapper around `File#open` and the expectation is the block then repeatedly calls `Report#report_saved` and `Report#report_deleted`

File: lib/datasets/report.rb [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    def volumes_to_s(volumes)
+      volumes.map{|volume| volume_to_s(volume) }
+        .join("\n")
+    end
+
+    def volume_to_s(volume)
+      "#{volume.namespace}.#{volume.id}"
+    end
+
+    attr_reader :fs
```
Comment (grosscol) [inline lib/datasets/report.rb]: move to top with other attrs

**Comment by malakai97** [comment]
> I thought the plan was to run it the first time with the entire repository (15 million volumes)
Indeed, but there are a couple ways we could accomplish that: 
1. Run this once, on a time period from epoch until now.  
2. Run this lots of times, on time periods of a month or so.

> VolumeActionLogger is super cheesy and the absolute minimum we would need to do, this is a substantial improvement, so fine to just dump it
We'll get it when we do some cleanup / reconciliation passes.

> code
I'll look at this. 

---

If the contents of this PR at present are sufficient, I'd suggest we merge them, and consider a possible refactor of the report interface in another issue with your code snippet.

File: spec/filesystem_spec.rb [hunk 2]
```diff
@@ -35,6 +35,36 @@ module Datasets
       end
       after(:all) { FileUtils.remove_entry_secure TMPPATH }
 
+      describe "#read" do
```
Comment (grosscol) [inline spec/filesystem_spec.rb]: Seems like we're testing the underlying standard library.

File: lib/datasets/report_manager.rb [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    # yields a new time range from the time the last report ended
+    # until now.  It then creates and saves a new report with the
+    # result of the block.
+    # @yieldparam new_range [Range<Time>] The period from when the last
+    #   report started until now.
+    # @yieldreturn [Array<Array<Volume>>] The block should return the
+    #   array of volumes that were saved and the array of volumes that
+    #   were removed.
+    # @return [Report]
+    def build_next_report(&block)
```
Comment (grosscol) [inline lib/datasets/report_manager.rb]: The use of this is 
```
ReportManager.new('f/foo/bar', my_filesys).build_next_report do 
  # Something that ruturns [saved_volumes, deleted_volumes]
end
```
That's going to get called from the scheduler?  Will each scheduler being using it's own subdirectory?

**Comment by malakai97** [comment]
> report.rb: move to top with other attrs

It's private here.  

> filesystem_spec.rb: Seems like we're testing the underlying standard library.

We're testing our thin wrapper over the standard library.  We could just check that File was called correctly, but I think this is a better test.

> report_manager.rb: That's going to get called from the scheduler? 

No, the thing invoking the scheduler will call this.  Here's the full, WIP snippet:

```ruby
module Datasets
  class SchedulerJob < Job

    def perform
      Datasets::ReportManager.new(Datasets.config.report_dir).build_next_report do |time_range|
        scheduler = Scheduler.new(
          volume_repo: Datasets.config.volume_repo[profile],
          src_path_resolver: Datasets.config.src_path_resolver[profile],
          volume_writer: Datasets.config.volume_writer[profile],
          filter: Datasets.config.filter[profile],
          time_range: time_range
        )
        [scheduler.add, scheduler.delete]
      end
    end

  end
end
```

> report_manager.rb: Will each scheduler being using it's own subdirectory?

Each rights profile should use its own subdirectory, yes.

**Comment by grosscol** [comment]
We're going to wrap the schedulers in a job?  So the CLI will kick off a scheduler job which will schedule adds and deletes?  That could be a problem for reporting.  There is no garuntee that we don't have two scheduler jobs running concurrently.  In that case, we'd have two ReportManagers trying to write to the directory at the same time.

**Comment by aelkiss** [comment]
Even if it is wrapped in a job we shouldn't have more than one scheduler running per subset - as indicated in datasets_cli the intention was there be some logic to make sure the previous run had completely finished before starting another one. That isn't implemented yet though, I'll create an issue

**Comment by malakai97** [comment]
The cli should just create the scheduler and run it.  The systemd service will have jobs run at specified times that create and invoke a scheduler.

**Pull request merged**